providers/implementations/rands/drbg_ctr.c

```diff
@@ -728,32 +732,58 @@ static int drbg_ctr_set_ctx_params_locked(void *vctx, const OSSL_PARAM params[])
         propquery = (const char *)p->data;
     }
 
+    if ((p = OSSL_PARAM_locate_const(params,
+                                     OSSL_PROV_PARAM_CORE_PROV_NAME)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        if ((prov = ossl_provider_find(libctx,
+                                       (const char *)p->data, 1)) == NULL)
+            return 0;
+    }
+
     if ((p = OSSL_PARAM_locate_const(params, OSSL_DRBG_PARAM_CIPHER)) != NULL) {
         const char *base = (const char *)p->data;
         size_t ctr_str_len = sizeof("CTR") - 1;
         size_t ecb_str_len = sizeof("ECB") - 1;
 
         if (p->data_type != OSSL_PARAM_UTF8_STRING
-                || p->data_size < ctr_str_len)
+                || p->data_size < ctr_str_len) {
+            ossl_provider_free(prov);
             return 0;
+        }
         if (OPENSSL_strcasecmp("CTR", base + p->data_size - ctr_str_len) != 0) {
             ERR_raise(ERR_LIB_PROV, PROV_R_REQUIRE_CTR_MODE_CIPHER);
+            ossl_provider_free(prov);
             return 0;
         }
-        if ((ecb = OPENSSL_strndup(base, p->data_size)) == NULL)
+        if ((ecb = OPENSSL_strndup(base, p->data_size)) == NULL) {
+            ossl_provider_free(prov);
             return 0;
+        }
         strcpy(ecb + p->data_size - ecb_str_len, "ECB");
         EVP_CIPHER_free(ctr->cipher_ecb);
         EVP_CIPHER_free(ctr->cipher_ctr);
-        ctr->cipher_ctr = EVP_CIPHER_fetch(libctx, base, propquery);
-        ctr->cipher_ecb = EVP_CIPHER_fetch(libctx, ecb, propquery);
+        /*
+         * Try to fetch algorithms from our own provider code, fallback
+         * to generic fetch only if that fails
+         */
+        (void)ERR_set_mark();
+        ctr->cipher_ctr = evp_cipher_fetch_from_prov(prov, base, NULL);
+        if (ctr->cipher_ctr == NULL)
+            ctr->cipher_ctr = EVP_CIPHER_fetch(libctx, base, propquery);
+        ctr->cipher_ecb = evp_cipher_fetch_from_prov(prov, ecb, NULL);
+        if (ctr->cipher_ecb == NULL)
+            ctr->cipher_ecb = EVP_CIPHER_fetch(libctx, ecb, propquery);
+        (void)ERR_pop_to_mark();
```

Comment:
This is more correct so we do not lose errors if both fetches fail:

```suggestion
        (void)ERR_set_mark();
        ctr->cipher_ctr = evp_cipher_fetch_from_prov(prov, base, NULL);
        if (ctr->cipher_ctr == NULL) {
            (void)ERR_pop_to_mark();
            ctr->cipher_ctr = EVP_CIPHER_fetch(libctx, base, propquery);
        } else {
            (void)ERR_clear_last_mark();
        }
        (void)ERR_set_mark();
        ctr->cipher_ecb = evp_cipher_fetch_from_prov(prov, ecb, NULL);
        if (ctr->cipher_ecb == NULL) {
            (void)ERR_pop_to_mark();
            ctr->cipher_ecb = EVP_CIPHER_fetch(libctx, ecb, propquery);
        } else {
            (void)ERR_clear_last_mark();
        }
```

Reply:
Ok I am changing this and then for consistency changing the order on the other two DRBG so they operate in the same order (wrt pop vs clear).